File: src/main/java/org/openhds/mobile/fragment/DataSelectionFragment.java
package org.openhds.mobile.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.RelativeLayout;

import org.openhds.mobile.R;
import org.openhds.mobile.provider.DatabaseAdapter;
import org.openhds.mobile.repository.DataWrapper;
import org.openhds.mobile.utilities.MessageUtils;

import java.util.List;

import static org.openhds.mobile.utilities.LayoutUtils.configureTextWithPayload;
import static org.openhds.mobile.utilities.LayoutUtils.makeTextWithPayload;

public class DataSelectionFragment extends Fragment {

    private DataSelectionListener listener;
    private DataSelectionListAdapter dataWrapperAdapter;
    private ListView listView;

    public interface DataSelectionListener {
        void onDataSelected(DataWrapper data);
    }

    @Override
    public void onAttach(Context ctx) {
        super.onAttach(ctx);
        if (ctx instanceof DataSelectionListener) {
            this.listener = (DataSelectionListener)ctx;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        this.listener = null;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ViewGroup viewGroup = (ViewGroup) inflater.inflate(R.layout.data_selection_fragment, container, false);
        listView = (ListView) viewGroup.findViewById(R.id.data_fragment_listview);
        listView.setOnItemClickListener(new DataClickListener());
        registerForContextMenu(listView);
        return viewGroup;
    }

    public void populateData(List<DataWrapper> data) {
        dataWrapperAdapter = new DataSelectionListAdapter(getActivity(), R.layout.generic_list_item_white_text, data);
        listView.setAdapter(dataWrapperAdapter);
    }

    public void clearData() {
        if (dataWrapperAdapter != null) {
            dataWrapperAdapter.clear();
        }
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        if (v.getId() == R.id.data_fragment_listview) {
            getActivity().getMenuInflater().inflate(R.menu.data_selection_menu, menu);
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save_favorite:
                AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();
                DataWrapper selected = getItem(info.position);
                if (selected != null) {
                    Context ctx = getActivity();
                    DatabaseAdapter.getInstance(ctx).addFavorite(selected);
                    MessageUtils.showShortToast(ctx, R.string.saved_favorite);
                }
                return true;
        }
        return super.onContextItemSelected(item);
    }

    private class DataClickListener implements OnItemClickListener {
        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            if (listener != null) {
                DataWrapper selected = getItem(position);
                listener.onDataSelected(selected);
            }
        }
    }

    private class DataSelectionListAdapter extends ArrayAdapter<DataWrapper> {

        public DataSelectionListAdapter(Context context, int resource, List<DataWrapper> objects) {
            super(context, resource, objects);
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {

            DataWrapper dataWrapper = getItem(position);

            if (convertView == null) {
                convertView = makeTextWithPayload(getActivity(), dataWrapper.getName(), dataWrapper.getExtId(), dataWrapper.getName(),
                        null, null, R.drawable.data_selector, dataWrapper.getStringsPayload(), dataWrapper.getStringIdsPayload(), false);
            } else {
                configureTextWithPayload(getActivity(), (RelativeLayout) convertView, dataWrapper.getName(), dataWrapper.getExtId(),
                        dataWrapper.getStringsPayload(), dataWrapper.getStringIdsPayload(), false);
            }

            return convertView;
        }
    }

    private DataWrapper getItem(int position) {
        return dataWrapperAdapter.getItem(position);
    }
}

File: src/main/kotlin/org/cimsbioko/fragment/DataSelectionFragment.kt
package org.cimsbioko.fragment

import android.content.Context
import android.os.Bundle
import android.view.*
import android.view.ContextMenu.ContextMenuInfo
import android.widget.AdapterView
import android.widget.AdapterView.AdapterContextMenuInfo
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.ProgressBar
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.cimsbioko.R
import org.cimsbioko.databinding.DataSelectionFragmentBinding
import org.cimsbioko.databinding.GenericListItemBinding
import org.cimsbioko.model.HierarchyItem
import org.cimsbioko.navconfig.HierFormatter
import org.cimsbioko.provider.DatabaseAdapter
import org.cimsbioko.utilities.MessageUtils.showShortToast
import org.cimsbioko.utilities.configureText
import org.cimsbioko.utilities.makeText
import org.cimsbioko.utilities.toLevelIcon
import org.cimsbioko.viewmodel.NavModel

class DataSelectionFragment : Fragment() {

    private val model: NavModel by activityViewModels()

    private var progressBar: ProgressBar? = null
    private var listView: ListView? = null
    private var adapter: DataSelectionListAdapter? = null
    private var itemFormatter: HierFormatter? = null

    private var isLoading: Boolean
        get() = progressBar?.isVisible == true
        set(loading) {
            progressBar?.visibility = if (loading) View.VISIBLE else View.GONE
            listView?.visibility = if (loading) View.GONE else View.VISIBLE
        }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return DataSelectionFragmentBinding.inflate(inflater, container, false).also {
            listView = it.dataFragmentListview.also { listView ->
                listView.onItemClickListener = DataClickListener()
                registerForContextMenu(listView)
            }
            progressBar = it.progressBar
        }.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        lifecycleScope.launch {
            model.childItems.asStateFlow().collectLatest { result ->
                isLoading = when (result) {
                    NavModel.ChildItems.Loading -> true
                    is NavModel.ChildItems.Loaded -> {
                        model.childItemFormatter?.also { populateData(result.items, it) }
                        false
                    }
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        listView?.let { unregisterForContextMenu(it) }
        listView = null
        progressBar = null
    }

    private fun populateData(data: List<HierarchyItem>, formatter: HierFormatter) {
        itemFormatter = formatter
        adapter = DataSelectionListAdapter(requireContext(), R.layout.generic_list_item, data)
        listView?.adapter = adapter
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        if (v.id == R.id.data_fragment_listview) {
            requireActivity().menuInflater.inflate(R.menu.data_selection_menu, menu)
        }
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.save_favorite) {
            getItem((item.menuInfo as AdapterContextMenuInfo).position)?.let { selected ->
                DatabaseAdapter.addFavorite(selected.wrapped)
                showShortToast(requireContext(), R.string.saved_favorite)
            }
            return true
        }
        return super.onContextItemSelected(item)
    }

    private inner class DataClickListener : OnItemClickListener {
        override fun onItemClick(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
            lifecycleScope.launch { getItem(position)?.also { model.stepDown(it.wrapped) } }
        }
    }

    private inner class DataSelectionListAdapter(
            context: Context,
            resource: Int,
            objects: List<HierarchyItem>
    ) : ArrayAdapter<HierarchyItem>(context, resource, objects) {
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            return getItem(position).let { item ->
                val formatted = item?.let { itemFormatter?.formatItem(item) }
                (convertView?.let { GenericListItemBinding.bind(it) }
                        ?: makeText(requireActivity(), layoutTag = formatted?.heading, background = R.drawable.data_selector)).apply {
                    configureText(requireActivity(),
                            text1 = formatted?.heading,
                            text2 = formatted?.subheading,
                            details = formatted?.details,
                            centerText = false,
                            iconRes = item?.level?.toLevelIcon(),
                            detailsPadding = resources.getDimensionPixelSize(R.dimen.detail_padding)
                    )
                }
            }.root
        }
    }

    private fun getItem(position: Int): HierarchyItem? {
        return adapter?.getItem(position)
    }
}